File: codes/mc-explorer/motif-master/algorithm/Graph.cpp
//
// Created by Hujiafeng on 16/10/2017.
//

#include"Graph.h"
#include"global-variable.h"


Graph::Graph() {
}

void Graph::Read(string path) {
	cout << "Start to load undirected graph from " << path << " for the basic subgraph matching algorithm." << endl;
	FILE *fin;
	fin = fopen(path.c_str(), "rb");
	if (fin == NULL) {
		printf("ERROR: network file not found!\n");
		exit(1);
	}

	fscanf(fin, "%d %d", &nodeNum, &edgeNum);
	cout << "#Nodes:" << nodeNum << " #Edegs:" << edgeNum << endl;

	int from, to;
	adjList.resize(nodeNum);

	if (directed) {
		InAdjList.resize(nodeNum);
		OutAdjList.resize(nodeNum);
	}

	int self_edge = 0;
	for (int i = 0; i < edgeNum; ++i) {
		fscanf(fin, "%d %d", &from, &to);

		if (CHECKSPECIES) {
			if (validNodes.find(from) == validNodes.end() || validNodes.find(to) == validNodes.end()) {
				continue;
			}
		}


		// Ignore self-edge
		if (to == from) {
			self_edge++;
			continue;
		}

		if (adjList[from].find(to) != adjList[from].end()) {
			// This edge has existed.
			continue;
		}
		adjList[from].insert(to);
		edges.push_back(EDGE(from, to));

		if (!directed) {
			// insert its reverse edge
			edges.push_back(EDGE(to, from));
			adjList[to].insert(from);
		} else {
			InAdjList[to].insert(from);
			OutAdjList[from].insert(to);
		}
	}
	int label;
	for (int i = 0; i < nodeNum; ++i) {
		fscanf(fin, "%d", &label);
#ifdef UNLABELED
		// Ignore all node labels.
			label=0;
#endif
		labels.push_back(label);
	}

	assert(labels.size() == nodeNum);
	edgeNum = edges.size();
	fclose(fin);
}

/*
 * print out the whole graph. [For debug]
 */
void Graph::Print() const {
	cout << "nid\tlabel\tNeighbors" << endl;
	for (int i = 0; i < nodeNum; i++) {
		cout << i << "\t" << labels[i] << "\t";
		for (const auto &e : adjList[i]) {
			cout << "\t" << e;
		}
		cout << endl;
	}
}

void Graph::Print(ostream &ofile) const {
	ofile << nodeNum << "\t" << edges.size() / 2 << endl;
	for (int i = 0; i < edges.size(); ++i) {
		if (edges[i].from < edges[i].to)
			ofile << edges[i].from << "\t" << edges[i].to << endl;
	}
	for (int i = 0; i < nodeNum; i++) {
		if (i == 0) ofile << labels[i];
		else ofile << "\t" << labels[i];
	}
	ofile << endl;
}

void Graph::clear() {
	for (auto &e : adjList) {
		e.clear();
	}
	adjList.clear();

	if (directed) {
		for (auto &e : InAdjList) {
			e.clear();
		}

		for (auto &e : OutAdjList) {
			e.clear();
		}
		InAdjList.clear();
		OutAdjList.clear();
	}

	edges.clear();
	labels.clear();
}

File: codes/java/src/main/java/mmc/Graph.java
package mmc;

import org.apache.log4j.Logger;

import java.util.HashSet;
import java.util.Vector;

public class Graph {
	private static Logger logger = Logger.getLogger(Graph.class);
	//Number of nodes and edges
	public int nodeNum;
	public int edgeNum;
	
	// The adjacency list of the graph; adjList[i] stores the neighbor set of node i
	public Vector<HashSet<Integer>> adjList = new Vector<HashSet<Integer>>();
	
	// The edge list
	public Vector<Edge> edges = new Vector<Edge>();
	public Vector<Edge> originalEdges = new Vector<Edge>();
	
	// labels of nodes. Node id {0, ..., nodeNum-1}
	public Vector<Integer> labels = new Vector<Integer>();
	
	public Graph(){};
	
	/*
     * Read the graph from the given path
     * Graph file format [all values are integers][undirected]:
     * Line1: n(nodeNum) m(edgeNum)
     * Line2: from_1 to_1
     * Line3: from_2 to_2
     * ...
     * Line_(m+1): from_m to_m
     * Line_(m+2): label of node_1
     * ...
     * Line_(1+m+n): label of node_n
     */
    void read(String path){
    	logger.info("Start to load undirected graph from " + path + " for the basic subgraph matching algorithm.");
    	
    };

    /*
     * print out the whole graph. [For debug]
     */
    public void print(){
    	 logger.info("nid\tlabel\tNeighbors");
    	    for (int i = 0; i < nodeNum; i++) {
    	        System.out.print(i + "\t" + labels.get(i));
    	        for(Integer e : adjList.get(i)) {
    	        	System.out.print("\t" + e);
    	        }
    	        logger.info("");
    	    }
    };
//
//    void Print(ostream &ofile){};

    public void clear(){
    	adjList.clear();
    	edges.clear();
    	labels.clear();
    };
}
